packages/ui/react-ui-editor/src/extensions/automerge/sync.ts

```diff
@@ -12,67 +12,59 @@ import { next as A } from '@dxos/automerge/automerge';
 import {
   getLastHeads,
   getPath,
-  isReconcile,
-  reconcileAnnotation,
-  updateHeads,
   type IDocHandle,
   type State,
+  reconcileAnnotation,
+  updateHeads,
+  isReconcile,
 } from './defs';
 import { updateAutomerge } from './update-automerge';
 import { updateCodeMirror } from './update-codemirror';
 
 /**
  * Implements three-way merge (on each mutation).
  */
-export class PatchSemaphore {
-  private _inReconcile = false;
+export class Syncer {
+  private _pending = false;
 
   // prettier-ignore
   constructor(
     private readonly _handle: IDocHandle,
     private readonly _state: StateField<State>
   ) {}
 
-  reconcile(view: EditorView) {
-    if (!this._inReconcile) {
-      this._inReconcile = true;
-      this.doReconcile(view);
-      this._inReconcile = false;
+  reconcile(view: EditorView, editor: boolean) {
+    if (this._pending) {
+      return;
     }
-  }
-
-  private doReconcile(view: EditorView) {
-    const path = getPath(view.state, this._state);
 
-    // Get the heads before the unreconciled transactions are applied.
-    const oldHeads = getLastHeads(view.state, this._state);
-    let selection = view.state.selection;
+    this._pending = true;
+    if (editor) {
+      this.onEditorChange(view);
+    } else {
+      this.onAutomergeChange(view);
+    }
+    this._pending = false;
+  }
 
-    // First undo all the unreconciled transactions.
+  onEditorChange(view: EditorView) {
     const transactions = view.state.field(this._state).unreconciledTransactions.filter((tx) => !isReconcile(tx));
-    const toInvert = transactions.slice().reverse();
-    for (const tr of toInvert) {
-      const inverted = tr.changes.invert(tr.startState.doc);
-      selection = selection.map(inverted);
+    const newHeads = updateAutomerge(this._state, this._handle, transactions, view.state);
+    if (newHeads) {
       view.dispatch({
-        changes: inverted,
-        annotations: reconcileAnnotation.of(true),
+        effects: updateHeads(newHeads),
+        annotations: reconcileAnnotation.of(false),
       });
     }
+  }
 
-    // Apply the unreconciled transactions to the document.
```

Comment:
Having the equivalent of some of these comments on the new code would be helpful for parsing what each of these methods is doing